docs/decisions/0001-directory-structure.md

```diff
@@ -0,0 +1,70 @@
+# Directory Structure
+
+## Summary
+
+We want to formalize the file and directory structure for the `cwms-python` module source code.
+
+## Considered Options
+
+Each top-level CWMS Data API endpoint corresponds to a `cwms-python` submodule. Nested endpoints would then map to a sub-directory within the submodule. The table below lists some example API endpoints and the corresponding submodule.
+
+| API Endpoint                                  | CWMS Python Module           |
+|-----------------------------------------------|------------------------------|
+| `/cwms-data/location/category`                | `cwms.location.category`     |
+| `/cwms-data/location/group`                   | `cwms.location.group`        |
+| `/cwms-data/locations`                        | `cwms.locations`             |
+| `/cwms-data/levels`                           | `cwms.levels`                |
+| `/cwms-data/specified-levels`                 | `cwms.specified_levels`      |
+| `/cwms-data/timeseries`                       | `cwms.timeseries`            |
+| `/cwms-data/timeseries/category`              | `cwms.timeseries.category`   |
+| `/cwms-data/timeseries/identifier-descriptor` | `cwms.timeseries.identifier` |
+| `/cwms-data/timeseries/group`                 | `cwms.timeseries.group`      |
+
+### Special Cases
+
+The direct mapping outlined above works well for most of the API endpoints. However, there are a couple cases which may warrant special consideration.
+
+#### Location Related Endpoints
+
+The locations, location categories, and location groups API endpoints return related data, despite having different URL paths (`/cwms-data/locations` vs. `/cwms-data/location/{category,group})`). From a `cwms-python` perspective, it may make more sense (and be less confusing to the user) if the corresponding functions were defined in the same submodule. This could be either `cwms.location` or `cwms.locations`, but since we are using the plural case for other endpoints (`cwms.levels`, `cwms-timeseries`, etc), `cwms.locations` would be the more natural choice.
+
+#### Levels and Specified Levels
+
+This case is similar to the one above. The `/cwms-data/specified-level` endpoint returns data that is related to levels (returned from `/cwms-data/levels`). Again, it makes sense to organize these function in a single submodule. There are two possible approaches:
```

Comment:
To avoid this confusion, we generally refer to the levels as "location levels" so I'd think `cwms.levels.location_levels` makes sense. The added characters at this child node level won't matter as the user will likely be importing `cwms.levels` anyways.

Reply:
This makes sense. My only question is: is the use of "location levels" (and "physical locations") common in both the CDA and cwms-python projects?

I only ask because the original intention (which admittedly the proposed ADR fails to accomplish 100%) was to have a one-to-one mapping between CDA endpoints and the cwms-python source files. For example, `/cwms-data/some-resource/some-sub-group/{resource-id}` would correspond to `cwms.some_resource.some_sub_group.py`. This way, a cwms-python developer would be able to immediately reference the CDA documentation related to the module they are working on, and a CDA developer who wants to look at the python implementation for a collection of endpoints would know exactly where to look, even if they are not a cwms-python developer.

In the current version of the proposal the CDA endpoints would map to the following modules:
- `/cwms-data/levels => cwms.levels.levels`
- `/cwms-data/locations => cwms.locations.locations`

If I understand your suggestions correctly these would be changed to:
- `/cwms-data/levels => cwms.levels.location_levels`
- `/cwms-data/locations => cwms.locations.physical_locations`

Neither is a direct mapping, but as someone who is new to the project, codebase, etc, the first seems a litte more intuitive (this may just be a personal preference). In the second case, I think I would get to the correct python file pretty quickly, but might have more difficulty finding the CDA documentation for the relevant endpoints. If both CDA and cwms-python developers commonly refer to "location levels" and "physical locations" then this is really only an issue for new developers who aren't familiar with the terminology. 

Since I don't have a lot of experience with the CWMS platform yet, my goal is to try to find ways to (hopefully) reduce the friction for potential future developers in the most general ways possible. I appreciate your experience and perspective when it comes to CWMS and if you think this is a clearer naming convention, I'll go ahead and update the ADR and make the changes listed above when renaming the modules.

Reply:
I wasn't involved in the CDA naming for `levels` and I would have advocated for `location-levels` as that is the standard CWMS naming. Additionally, I find `cwms.levels.levels` to be confusing as a python user (might just be me).

The locations one is even more confusing given that `location/groups/` and `location/categories/` endpoints differ from the `locations/` endpoint. I think from the python perspective, combining them makes sense, though we also have to think forward to the additional endpoints that will be added. There will be a `outlets/`, `turbines/`, `pumps/`, `projects/`, etc endpoints added. They are all locations and could fall under the `cwms.locations` module? 

`cwms.locations.outlets`
`cwms.locations.turbines`
`cwms.locations.projects`
`cwms.locations.pumps`

I don't know what the CDA endpoints will look like for these data types. Maybe:

`/outlets`
`/outlets/gate-changes`
`/turbines/turbine-changes`
`/turbines`

Or if they will be endpoints off of the `location` endpoint like groups:

`/location/outlets`
`/location/outlets/gate-changes`
`/location/turbines/turbine-changes`
`/location/turbines`
`/location/turbines/turbine-changes`

Reply:
I agree that `cwms.levels.levels` is a bit awkward (same goes for `cwms.locations.locations`), but ultimately the API that we expose to the user would not require them to import from those submodules directly.

Here's my main concern and the issue that I'm attempting to address: at some point in the future, after interfaces for all the CDA endpoints have been added, we are going to have a fairly large codebase that will need to be updated and maintained. Possibly by new developers who, like me, may not have a deep understanding of CWMS/CDA. Navigating a large codebase with lots of modules, files, etc can be difficult, particularly when you are new to a project. If we have some convention that relates our source files in cwms-python to the CDA endpoints that they call, we can make it much easier for developers of both projects (particularly when onboarding new developers).

As an example from another field, Ember is a frontend framework that follows a strict convention when it comes to files and how they map to application routes. Once you know the convention, if there is an error on any page of even the most massive application, you would immediately know which file you need to look at. Without getting into the pros/cons of Ember as a framework, there is something really great about being able to jump into a large project and feel comfortable on day one.

I'm not suggesting that we adhere to something as strictly as Ember, but the hope for this ADR was that we could establish enough guidelines that new (or external) developers would be able to start contributing without having to spend a huge amount of effort figuring out how to navigate the codebase, docs, etc.

With the uncertainty around the outlets, turbines, and other endpoints, I'm not sure what's the best way to proceed. If those are implemented as top-level endpoints (like in the first case) but should be associated with locations in cwms-python, a fairly large share of submodules would be treated as exceptions to rule.

Reply:
I agree with jordan.  We should name the folder and files according to the api endpoints.  Inside the cwms.levels.levels.  the function will be called get_location_levels or something like that.  So when the user is actually using it they would call

import cwms

cwms.level.get_location_level.

or 

from cwms import levels

level.get_location_levels.

the user is only going to see the function and not the added level.levels.  But the developer can know that the functions in the levels.levels is associated with the levels - /levels endpoint.

Reply:
Yeah that's exactly the reason I'm advocating for `cwms.levels.location_levels` and `cwms.locations.physical_locations`. As a CWMS developer, those terms are standard and match database tables/views, but as a non-CWMS developer (i.e. the end python user) will never even see that last layer so it should avoid any confusion there. We do get into murky water adding python developers to this project that aren't as familiar with the CWMS backend, but documentation can help there.

That all being said, maybe I'm the only one who finds `cwms.levels.levels` and `cwms.locations.locations` confusing from a developer perspective. I was also not around for the naming/design of the `/levels` and `/locations`/ endpoints so maybe this discussion has already run its course by CDA devs.

Reply:
Thinking on this more:

physical locations are the parent set of metadata for all location types in the database, so I can see the generic `locations` module name being used as it is the base for all location data. I just want to make sure that we don't wind up with a developer putting a `retrieve_outlet` method in the `cwms.locations.locations` module since an outlet is a location. Less of a concern for levels since there aren't really hierarchies of levels (just types like constant, seasonal, timseries, virtual).

Reply:
I've updated the ADR to include the filename changes above and marked it as accepted, since it seems that we're all pretty much on the same page. Let me know if there are any additional changes that need to be made.